go/trace/trace.go

```diff
@@ -20,34 +20,51 @@ limitations under the License.
 package trace
 
 import (
-	"golang.org/x/net/context"
+  "flag"
+  "io"
+  "strings"
+
+  "github.com/opentracing/opentracing-go"
+  "golang.org/x/net/context"
+  "vitess.io/vitess/go/vt/log"
+  "vitess.io/vitess/go/vt/proto/vtrpc"
+  "vitess.io/vitess/go/vt/vterrors"
 )
 
 // Span represents a unit of work within a trace. After creating a Span with
 // NewSpan(), call one of the Start methods to mark the beginning of the work
 // represented by this Span. Call Finish() when that work is done to record the
 // Span. A Span may be reused by calling Start again.
 type Span interface {
-	// StartLocal marks the beginning of a span representing time spent doing
-	// work locally.
-	StartLocal(label string)
-	// StartClient marks the beginning of a span representing time spent acting as
-	// a client and waiting for a response.
-	StartClient(label string)
-	// StartServer marks the beginning of a span representing time spent doing
-	// work in service of a remote client request.
-	StartServer(label string)
-	// Finish marks the span as complete.
 	Finish()
 	// Annotate records a key/value pair associated with a Span. It should be
 	// called between Start and Finish.
 	Annotate(key string, value interface{})
 }
 
+type SpanType int
+
+const (
+  Local SpanType = iota
+  Client
+  Server
+)
+
 // NewSpan creates a new Span with the currently installed tracing plugin.
 // If no tracing plugin is installed, it returns a fake Span that does nothing.
-func NewSpan(parent Span) Span {
-	return spanFactory.New(parent)
+func NewSpan(inCtx context.Context, label string, spanType SpanType) (Span, context.Context) {
+  parent, _ := spanFactory.FromContext(inCtx)
+  span := spanFactory.New(parent, label, spanType)
+  outCtx := spanFactory.NewContext(inCtx, span)
+
+  return span, outCtx
+}
+
+// NewClientSpan returns a span and a context to register calls to dependent services
+func NewClientSpan(inCtx context.Context, serviceName, spanLabel string) (Span, context.Context) {
+  span, ctx := NewSpan(inCtx, spanLabel, Client)
```

Comment:
This is in place so that services we are calling out to that do not themselves support tracing are still traced as such. "peer.service" is not a jaeger thing, it's an OpenTracing thing: https://github.com/opentracing/specification/blob/master/semantic_conventions.md